### ISSUE TITLE
`__bool__` overloading not detected for narrowing when object type is masked by supertype

### ISSUE THREAD

**Issue by kwshi** [microsoft/pyright#9577]
**Describe the bug**
In general, instances of user-defined classes are truthy by default unless the class overloads `__bool__`. Pyright knows this and uses it for narrowing: e.g.,

```python
class A:
    pass

class B:
    x: int
    def __init__(self):
        self.x = int(input())
    def __bool__(self):
        return self.x > 5

def f(a: A | None):
    if not a:
        typing.reveal_type(a) # only `None` because `A` instances are always truthy

def g(b: B | None):
    if not b:
        typing.reveal_type(b) # still `B | None` because `B` instances are sometimes falsy as well
```

This detection slips through the gaps if instances of `B` "masquerade" as some different (but still valid) type, e.g. subtyping. Suppose we actually defined `B` to be a subclass of `A`; then it becomes legal to pass instances of `B` into `f`, but optional falsiness is _not_ detected in narrowing.

This sort of thing happens for other (implicit) subtype-y things too, e.g. `f` expects something satisfying a protocol and `B` satisfies that protocol. In fact this is the type of scenario in which I ran into this issue.

**Code or Screenshots**

Here's a more direct example:

```python
# pyright: strict
import typing

class A:
    pass

class B(A):
    def __bool__(self):
        return False

def f(a: A | None):
    if not a:
        typing.reveal_type(a)  # pyright says `None`...

f(B()) # but running this actually prints `Runtime type is 'B'`
```

**VS Code extension or command-line**
pyright 1.1.382, via CLI as well as neovim language server

**Other comments**

So, in theory the only sound way to address this is to get rid of boolean narrowing altogether, because literally any type can be subtyped with something that overloads `__bool__` in arbitrary ways. That's maybe a hefty price to pay; I think it's fine to not bother, but if so it might be nice for the documentation to acknowledge somewhere that 100% soundness is not guaranteed with examples like this one to show when soundness is violated, and/or specific assumptions needed to ensure soundness.

As an aside, mypy does not appear to do boolean narrowing; in the original example mypy always reveals type to be `* | None`. I think this is one of the differences listed in https://microsoft.github.io/pyright/#/mypy-comparison?id=type-guards. Looking at that section, I'd guess that overloading `__eq__` for literal types (e.g., subclassing `int`) might also result in similar narrowing "mistakes".

**Comment by erictraut** [comment]
Yes, this is a case where a type checker needs to make an assumption that may not be sound 100% of the time. Practical tradeoffs like this are required to make a static type checker usable.
